### PR TITLE
Added a motor proposal

### PR DESCRIPTION
**Pull request by omarhogni**

### CODE REVIEW

File: docs/functionality/motors.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# Motors interface under tfc
+## SpeedRatio
+A concept that
+maps a motors capable
+range from 1-100%
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:5]: should we mention the minus range here as well?

File: docs/functionality/motors.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+of frequencies, rpm's and gear
+ratios in our motors, causing
+many other physical quantities
+to be a bad fit for simple general
+purpose use cases.
+
+## Activating a running motor
+If activating a running motor
+the motor shall run with the new
+parameters provided, fe. changing
```
Comment (magni-mar) [inline docs/functionality/motors.md:27]: e.g. not fe.

File: docs/functionality/motors.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+### Stopping
+ (-1<=>1) is outside the motors
+ range by definition and
+ the motor shall be stopped
+ if a speed is applied in that range.
+
+### Rationalization
+We have a plethora of
+use cases where precise motor
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:15]:
```suggestion
use cases where explicit motor
```

File: docs/functionality/motors.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+also has a decent distance estimate, an 
+async motor coupled with a tachometer or an
+encoder has even better precision.
+
+
+
+## Interface idea
+```cpp
+int main(){
+    motor_types m{};
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:72]:
```suggestion
    motor_interface m{};
```
?

File: docs/functionality/motors.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    /// Liquid transport
+    m.pump(50%);
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:75]:
```suggestion
```

File: docs/functionality/motors.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    m.pump(10 l / s);
+    m.pump(10 l / min, 10 l, [](const std::error_code&){});
+    m.pump(10 l / min, 10 minutes, [](const std::error_code&){});
+
+    /// Linear transport
+    m.convey(50%);
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:81]:
```suggestion
```

File: docs/functionality/motors.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+    // Absolute positioning
+    // 0                               100
+    // |               <=>             |
+    m.move(50%); // NOT SPEEDRATIO
+    m.move(10m);
+    m.move(10m, [](const std::error_code&){}););
+
+    /// Rotational transport
+    m.rotate(50%);
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:96]:
```suggestion
```

File: docs/functionality/motors.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    m.notify(10l, ()[]{});
+    m.notify(10m, ()[]{});
+    m.notify(10 * 2*pi, ()[]{});
+
+
+    /// Special cases
+    m.run(0%);  // Stop the motor
+    m.run(0m/s);  // Stop the motor
+    m.run(0 * 2*pi);  // Stop the motor
+    m.run(); // Start the motor at configured speed
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:129]:
```suggestion
    m.run(); // Start the motor at configured speed
```

File: docs/functionality/motors.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    m.convey(10m/s);
+    m.convey(10m/s, 10meters, [](const std::error_code&){});
+    m.convey(10m/s, 10 minutes, [](const std::error_code&){});
+    m.convey(10 meters, [](const std::error_code&){});
+    m.convey(10 minutes, [](const std::error_code&){});
+
+    // Absolute positioning
+    // 0                               100
+    // |               <=>             |
+    m.move(50%); // NOT SPEEDRATIO
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:91]:
```suggestion
```
causes confusion

File: docs/functionality/motors.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+
+## Interface idea
+```cpp
+int main(){
+    motor_types m{};
+
+    /// Liquid transport
+    m.pump(50%);
+    m.pump(10 l / s);
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:76]:
```suggestion
    m.pump(); // config speed
    m.pump(10 l / s);
```

File: docs/functionality/motors.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    // If a sequence of
+    // m.run(50%);
+    // m.stop(100ms);
+    // m.run(50%);
+    // m.stop(); <- This uses the configured deceleration
+    m.stop(100ms);
+    m.quick_stop(); // Stop the motor with quick stop
+
+    /// Getting notified of a continously running motor
+    m.run(10l/min);
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:119]:
```suggestion
```

File: docs/functionality/motors.md (hunk 4 above)
Comment (jbbjarnason) [reply docs/functionality/motors.md:72]: Could be less confusing
general_interface
pump_interface : general_interface
convey_interface : general_interface
rotate_interface : general_interface

**Review by jbbjarnason — APPROVED** [review]

File: docs/functionality/motors.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    m.convey(10m/s, 10 minutes, [](const std::error_code&){});
+    m.convey(10 meters, [](const std::error_code&){});
+    m.convey(10 minutes, [](const std::error_code&){});
+
+    // Absolute positioning
+    // 0                               100
+    // |               <=>             |
+    m.move(50%); // NOT SPEEDRATIO
+    m.move(10m);
+    m.move(10m, [](const std::error_code&){}););
```
Comment (jbbjarnason) [inline docs/functionality/motors.md:93]:
```suggestion
    m.move(10m, [](const std::error_code&){}););
    m.move_home();
```

File: docs/functionality/motors.md (hunk 3 above)
Comment (jbbjarnason) [reply docs/functionality/motors.md:19]: To my understanding 'precise' is deduced from precision and the control of the motor is still precise even though the speed is set relative to lsp-hsp.
So I suggested 'explicit', meaning that you don't need to explicitly set the velocity (m/s) as it is relative.

**Comment by jbbjarnason** [comment]
As a follow up. Maybe a general daemon should be made with vector of interfaces, so if no business logic is controlling a motor the Daemon could run() the motor at the configured run speed when running operation mode is entered, and call run(config->clean_speed) on cleaning mode

**Pull request merged**